phys2bids/slice4phys.py

```diff
@@ -106,6 +109,123 @@ def find_runs(phys_in, ntp_list, tr_list, thr=None, padding=9):
+        the BIDS dataset (True) or to leave it as-is (False). Default is False.
+
+    Returns
+    -------
+    phys_in_slices : dict
+        Each key is a run-wise filename (possibly further split by frequency)
+        and each value is a BlueprintInput object.
+
+    Notes
+    -----
+    The goal of this function is to abstract out the general slicing procedure
+    from slice4phys.
+    """
+    phys = deepcopy(phys)
+    if update_trigger:
+        phys.freq.append(phys.freq[phys.trigger_idx])
+        phys.units.append(phys.units[phys.trigger_idx])
+        phys.timeseries.append(phys.timeseries[phys.trigger_idx].copy())
+        phys.ch_name.append('original trigger')
+
+        # Fix up the trigger time series
+        phys.timeseries[phys.trigger_idx][:] = 0
+        for ons, off in run_timestamps.values():
+            phys.timeseries[phys.trigger_idx][ons:off] = 1
+
+    if not isinstance(padding, tuple):
+        time_before, time_after = padding, padding
+    else:
+        time_before, time_after = padding
+
+    phys_in_slices = {}
+    for i_run, fname in enumerate(run_timestamps.keys()):
+        run_attributes = run_timestamps[fname]  # tmp variable to collect run's info
+        trigger_onset, trigger_offset = run_attributes
+        min_onset, max_offset = 0, phys.timeseries[phys.trigger_idx].shape[0]
+
+        unique_frequencies = np.unique(phys.freq)
+        trigger_freq = phys.freq[phys.trigger_idx]
+
+        # Limit padding based on beginning and end of physio recording.
+        # We could also limit padding to prevent overlap between scans, if desired.
+        run_time_before = np.minimum(time_before, (trigger_onset - min_onset) / trigger_freq)
+        run_time_after = np.minimum(time_after, (max_offset - trigger_offset) / trigger_freq)
+
+        for freq in unique_frequencies:
+            to_subtract = int(run_time_before * freq)
+            to_add = int(run_time_after * freq)
+
+            run_onset_idx = int(trigger_onset * freq / trigger_freq) - to_subtract
+            run_offset_idx = int(trigger_offset * freq / trigger_freq) + to_add
+
+            # Split into frequency-specific object limited to onset-offset
+            temp_phys_in = deepcopy(phys[run_onset_idx:run_offset_idx])
+
+            if temp_phys_in.freq[temp_phys_in.trigger_idx] != freq:
+                example_ts = phys.timeseries[phys.freq.index(freq)]
+
+                # Determine time
+                new_time = (np.arange(example_ts.shape[0]) / temp_phys_in.timeseries[
+                    phys.freq.index(freq)].freq
+                )
+                new_time += np.min(temp_phys_in.timeseries[0])
+                temp_phys_in.timeseries[0] = new_time
+                temp_phys_in.freq[0] = freq
+
+                # Resample trigger
+                new_trigger = temp_phys_in.timeseries[temp_phys_in.trigger_idx]
+                new_trigger = resample(new_trigger, example_ts.shape[0], window=10)
```

Comment:
That is an interesting solution. I think I just took my original approach from a StackOverflow post, so any more tailored advice is definitely appreciated. I've switched over to pure `numpy`, and I think it will probably be fine since `phys2bids` uses `numpy` on trigger channels anyway.